sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py

```diff
@@ -4,150 +4,401 @@
+                sample_idx, first_feature_idx, X_t, centroids_window, dot_products
+            )
+        elif is_last_centroid_window:
+            accumulate_last_centroid_window_dot_products(
+                sample_idx, first_feature_idx, X_t, centroids_window, dot_products
+            )
+        else:
+            accumulate_full_window_dot_products(
+                sample_idx, first_feature_idx, X_t, centroids_window, dot_products
+            )
 
-    return _initialize_window_of_centroids, _load_window_of_centroids_and_features
+    return (
+        initialize_window_of_centroids,
+        load_window_of_centroids_and_features,
+        accumulate_dot_products,
+    )
 
 
-def _make_accumulate_sum_of_ops_kernel_func(
-    n_samples, n_features, window_n_features, window_n_centroids, ops, dtype
-):
+class _InitializeWindowKernelFuncFactory:
+    def __init__(
+        self,
+        dtype,
+        initialize_window_of_centroids_half_l2_norms,
+    ):
+        self.dtype = dtype
+        self.initialize_window_of_centroids_half_l2_norms = (
+            initialize_window_of_centroids_half_l2_norms
+        )
+
+    def make(self, window_n_centroids):
+        zero = self.dtype(0.0)
+
+        @dpex.func
+        def _initialize_results(results):
+            # Initialize the partial pseudo inertia dot product for each
+            # of the window_n_centroids centroids in the window.
+            for i in range(window_n_centroids):
+                results[i] = zero
+
+        if not self.initialize_window_of_centroids_half_l2_norms:
+            return _initialize_results
+
+        @dpex.func
+        # fmt: off
+        def _initialize_window_of_centroids(
+            local_work_id,                      # PARAM
+            first_centroid_idx,                 # PARAM
+            centroids_half_l2_norm,             # IN
+            window_of_centroids_half_l2_norms,  # OUT
+            results,                            # OUT
+        ):
+        # fmt: on
+            _initialize_results(results)
 
-    zero = dtype(0.0)
+            # The first `window_n_centroids` work items cooperate on loading the
+            # values of centroids_half_l2_norm relevant to current window. Each work
+            # item loads one single value.
+            if local_work_id < window_n_centroids:
+                half_l2_norm_loading_idx = first_centroid_idx + local_work_id
+                window_of_centroids_half_l2_norms[local_work_id] = centroids_half_l2_norm[
+                    half_l2_norm_loading_idx]
 
-    accumulate_dot_product = ops == "product"
-    accumulate_squared_diff = ops == "squared_diff"
+        return _initialize_window_of_centroids
 
-    if not accumulate_dot_product and not accumulate_squared_diff:
-        raise ValueError(
-            f'Expected ops to take values "product" or "squared_diff", got "{ops}" '
-            "instead."
+
+class _LoadWindowKernelFuncsFactory:
```

Comment:
What is the reason for preferring using using instances of classes (`_LoadWindowKernelFuncsFactory` and `_AccumulateSumOfOpsKernelFuncFactory` rather than just `(_)make_`-prefixed functions? Is this needed for some storing some state?

Reply:
I have the same question.